File: typescript/packages/core/src/commands/builtin/mongodb/grep.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { MongoDBAccessor } from '../../../accessor/mongodb.ts'
import type { IndexCacheStore } from '../../../cache/index/store.ts'
import { listDatabases } from '../../../core/mongodb/_client.ts'
import { resolveGlob } from '../../../core/mongodb/glob.ts'
import { read as mongoRead } from '../../../core/mongodb/read.ts'
import { readdir as mongoReaddir } from '../../../core/mongodb/readdir.ts'
import { detectScope } from '../../../core/mongodb/scope.ts'
import {
  formatGrepResults,
  searchCollection,
  searchDatabase,
} from '../../../core/mongodb/search.ts'
import { stat as mongoStat } from '../../../core/mongodb/stat.ts'
import { ScopeLevel } from '../../../core/mongodb/types.ts'
import { IOResult } from '../../../io/types.ts'
import { type FileStat, type PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { grepGeneric } from '../generic/grep.ts'
import { patternArg } from '../grep_helper.ts'
import { formatRecords } from '../utils/output.ts'
import { fileReadProvision } from './_provision.ts'

async function* mongoStream(
  accessor: MongoDBAccessor,
  p: PathSpec,
  index?: IndexCacheStore,
): AsyncIterable<Uint8Array> {
  yield await mongoRead(accessor, p, index)
}

async function grepCommand(
  accessor: MongoDBAccessor,
  paths: PathSpec[],
  texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  const pattern = patternArg(texts, opts.flags)
  const limit = accessor.config.defaultSearchLimit

  const first = paths[0]
  if (first !== undefined && pattern !== null) {
    const scope = detectScope(first)

    if (scope.level === ScopeLevel.ROOT) {
      const dbs = await listDatabases(accessor)
      const results: Awaited<ReturnType<typeof searchDatabase>> = []
      for (const db of dbs) {
        results.push(...(await searchDatabase(accessor, db, pattern, limit)))
      }
      const allLines = formatGrepResults(results)
      if (allLines.length === 0) return [new Uint8Array(0), new IOResult({ exitCode: 1 })]
      return [formatRecords(allLines), new IOResult()]
    }

    if (scope.level === ScopeLevel.DATABASE && scope.database !== null) {
      const results = await searchDatabase(accessor, scope.database, pattern, limit)
      const allLines = formatGrepResults(results)
      if (allLines.length === 0) return [new Uint8Array(0), new IOResult({ exitCode: 1 })]
      return [formatRecords(allLines), new IOResult()]
    }

    if (scope.level === ScopeLevel.ENTITY && scope.database !== null && scope.name !== null) {
      const docs = await searchCollection(accessor, scope.database, scope.name, pattern, limit)
      if (docs.length === 0) return [new Uint8Array(0), new IOResult({ exitCode: 1 })]
      const results = [{ database: scope.database, collection: scope.name, docs }]
      const allLines = formatGrepResults(results)
      return [formatRecords(allLines), new IOResult()]
    }
  }

  const resolved =
    paths.length > 0 ? await resolveGlob(accessor, paths, opts.index ?? undefined) : []
  const stat = (p: PathSpec): Promise<FileStat> => mongoStat(accessor, p, opts.index ?? undefined)
  const readdir = (p: PathSpec): Promise<string[]> =>
    mongoReaddir(accessor, p, opts.index ?? undefined)
  return grepGeneric('grep', resolved, texts, opts, stat, readdir, (p) =>
    mongoStream(accessor, p, opts.index ?? undefined),
  )
}

export const MONGODB_GREP = command({
  name: 'grep',
  resource: ResourceName.MONGODB,
  spec: specOf('grep'),
  fn: grepCommand,
  provision: fileReadProvision,
})

File: python/mirage/commands/builtin/mongodb/grep.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
from collections.abc import AsyncIterator

from mirage.accessor.mongodb import MongoDBAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.generic.grep import grep as generic_grep
from mirage.commands.builtin.grep_helper import pattern_arg
from mirage.commands.builtin.mongodb._provision import file_read_provision
from mirage.commands.builtin.utils.output import format_records
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.commands.spec.types import FlagView
from mirage.core.mongodb._client import list_databases
from mirage.core.mongodb.glob import resolve_glob
from mirage.core.mongodb.read import read as mongodb_read
from mirage.core.mongodb.readdir import readdir as _readdir
from mirage.core.mongodb.scope import detect_scope
from mirage.core.mongodb.search import (format_grep_results, search_collection,
                                        search_database)
from mirage.core.mongodb.stat import stat as _stat
from mirage.core.mongodb.stream import read_stream
from mirage.core.mongodb.types import ScopeLevel
from mirage.io.types import ByteSource, IOResult
from mirage.provision.types import ProvisionResult
from mirage.types import PathSpec


async def grep_provision(
    accessor: MongoDBAccessor,
    paths: list[PathSpec],
    *texts: str,
    **_extra: object,
) -> ProvisionResult:
    return await file_read_provision(
        accessor, paths,
        "grep " + " ".join(texts + tuple(str(p) for p in paths)))


@command("grep",
         resource="mongodb",
         spec=SPECS["grep"],
         provision=grep_provision)
async def grep(
    accessor: MongoDBAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: AsyncIterator[bytes] | bytes | None = None,
    prefix: str = "",
    index: IndexCacheStore = None,
    **flags: object,
) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(flags, spec=SPECS["grep"])
    pattern = pattern_arg(texts, fl)

    config = accessor.config
    limit = config.default_search_limit

    if paths and pattern is not None:
        scope = detect_scope(paths[0])

        if scope.level in (ScopeLevel.ENTITY, ScopeLevel.DATABASE,
                           ScopeLevel.ROOT):
            if scope.level != ScopeLevel.ROOT:
                await _stat(accessor, paths[0], index=index)
            entity_match = (scope.level == ScopeLevel.ENTITY and scope.database
                            and scope.name)
            if entity_match:
                docs = await search_collection(
                    accessor.client,
                    scope.database,
                    scope.name,
                    pattern,
                    limit=limit,
                )
                results = [(scope.database, scope.name, docs)] if docs else []
            elif scope.level == ScopeLevel.DATABASE and scope.database:
                results = await search_database(
                    accessor.client,
                    scope.database,
                    pattern,
                    limit=limit,
                )
            else:
                databases = await list_databases(
                    accessor.client,
                    config,
                )
                tasks = [
                    search_database(
                        accessor.client,
                        db_name,
                        pattern,
                        limit=limit,
                    ) for db_name in databases
                ]
                nested = await asyncio.gather(*tasks)
                results = []
                for r in nested:
                    results.extend(r)

            all_lines = format_grep_results(results)
            if not all_lines:
                return b"", IOResult(exit_code=1)
            return format_records(all_lines), IOResult()

    resolved = await resolve_glob(accessor, paths,
                                  index=index) if paths else []
    return await generic_grep(
        resolved,
        texts,
        flags,
        readdir=_readdir,
        stat=_stat,
        read_bytes=mongodb_read,
        read_stream=read_stream,
        accessor=accessor,
        stdin=stdin,
        index=index,
    )
